### ISSUE TITLE
Increase the usage of augmented assignment statements

### ISSUE THREAD

**Issue by elfring** [PyTorchLightning/pytorch-lightning#10512]
## Proposed refactoring or deprecation

:eyes: Some source code analysis tools can help to find opportunities for improving software components.
:thought_balloon: I propose to [increase the usage of augmented assignment statements](https://docs.python.org/3/reference/simple_stmts.html#augmented-assignment-statements "Augmented assignment statements") accordingly.

Would you like to integrate anything from a transformation result which can be generated by a command like the following?
(:point_right: Please check also for questionable change suggestions because of an evolving search pattern.)
```
lokal$ perl -p -i.orig -0777 -e 's/^(?<indentation>\s+)(?<target>\S+)\s*=\s*\k<target>[ \t]*(?<operator>[+\-%&|^@]|\*\*?|\/\/?|<<|>>)/$+{indentation}$+{target} $+{operator}=/gm' $(find ~/Projekte/pytorch-lightning/lokal -name '*.py')
```

### Additional context

cc @justusschock @awaelchli @akihironitta

**Comment by akihironitta** [comment]
@elfring Hi, thank you for your suggestion. Would you mind sharing the motivation for this change other than [PEP203](https://www.python.org/dev/peps/pep-0203/#rationale) which you've shared in other issues?
- https://github.com/pypa/pip/issues/10624
- https://github.com/nedbat/coveragepy/issues/1279

I believe it won't work properly at some places if we use augmented assignments as they are in-place operations.

- [What is `in-place operation`?](https://discuss.pytorch.org/t/what-is-in-place-operation/16244/13)
- [In-place Operations in PyTorch - What are they and why avoid them](https://towardsdatascience.com/in-place-operations-in-pytorch-f91d493e970e)

<details><summary>diff of this proposal</summary>

```diff
diff --git a/pytorch_lightning/callbacks/progress/rich_progress.py b/pytorch_lightning/callbacks/progress/rich_progress.py
index b07b48792..c58172e73 100644
--- a/pytorch_lightning/callbacks/progress/rich_progress.py
+++ b/pytorch_lightning/callbacks/progress/rich_progress.py
@@ -328,7 +328,7 @@ class RichProgressBar(ProgressBarBase):
         if total_train_batches != float("inf"):
             # val can be checked multiple times per epoch
             val_checks_per_epoch = total_train_batches // trainer.val_check_batch
-            total_val_batches = total_val_batches * val_checks_per_epoch
+            total_val_batches *= val_checks_per_epoch
 
         total_batches = total_train_batches + total_val_batches
 
diff --git a/pytorch_lightning/callbacks/progress/tqdm_progress.py b/pytorch_lightning/callbacks/progress/tqdm_progress.py
index 672d9d893..cbafd44d7 100644
--- a/pytorch_lightning/callbacks/progress/tqdm_progress.py
+++ b/pytorch_lightning/callbacks/progress/tqdm_progress.py
@@ -226,7 +226,7 @@ class TQDMProgressBar(ProgressBarBase):
         if total_train_batches != float("inf") and total_val_batches != float("inf"):
             # val can be checked multiple times per epoch
             val_checks_per_epoch = total_train_batches // trainer.val_check_batch
-            total_val_batches = total_val_batches * val_checks_per_epoch
+            total_val_batches *= val_checks_per_epoch
         total_batches = total_train_batches + total_val_batches
         reset(self.main_progress_bar, total=total_batches, current=self.train_batch_idx)
         self.main_progress_bar.set_description(f"Epoch {trainer.current_epoch}")
diff --git a/pytorch_lightning/loops/optimization/optimizer_loop.py b/pytorch_lightning/loops/optimization/optimizer_loop.py
index 7050ac75d..03dc2a148 100644
--- a/pytorch_lightning/loops/optimization/optimizer_loop.py
+++ b/pytorch_lightning/loops/optimization/optimizer_loop.py
@@ -87,7 +87,7 @@ class ClosureResult(OutputResult):
         if closure_loss is not None:
             # accumulate the loss. If ``accumulate_grad_batches == 1``, no effect
             # note: avoid in-place operation `x /= y` here on purpose
-            closure_loss = closure_loss / normalize
+            closure_loss /= normalize
 
         return cls(closure_loss, extra=extra)
 
diff --git a/pytorch_lightning/plugins/training_type/tpu_spawn.py b/pytorch_lightning/plugins/training_type/tpu_spawn.py
index 7aa4a6772..c6d885327 100644
--- a/pytorch_lightning/plugins/training_type/tpu_spawn.py
+++ b/pytorch_lightning/plugins/training_type/tpu_spawn.py
@@ -250,7 +250,7 @@ class TPUSpawnPlugin(DDPSpawnPlugin):
         output = xm.mesh_reduce("reduce", output, sum)
 
         if isinstance(reduce_op, str) and reduce_op.lower() in ("avg", "mean"):
-            output = output / self.world_size
+            output /= self.world_size
 
         return output
 
diff --git a/pytorch_lightning/trainer/supporters.py b/pytorch_lightning/trainer/supporters.py
index 6e2e51e82..0a18bf47a 100644
--- a/pytorch_lightning/trainer/supporters.py
+++ b/pytorch_lightning/trainer/supporters.py
@@ -86,7 +86,7 @@ class TensorRunningAccum:
         self.current_idx += 1
 
         # reset index when hit limit of tensor
-        self.current_idx = self.current_idx % self.window_length
+        self.current_idx %= self.window_length
         if self.current_idx == 0:
             self.rotated = True
 
diff --git a/pytorch_lightning/utilities/distributed.py b/pytorch_lightning/utilities/distributed.py
index 174051892..581a0cdc2 100644
--- a/pytorch_lightning/utilities/distributed.py
+++ b/pytorch_lightning/utilities/distributed.py
@@ -172,7 +172,7 @@ def sync_ddp(
     torch.distributed.all_reduce(result, op=op, group=group, async_op=False)
 
     if divide_by_world_size:
-        result = result / torch.distributed.get_world_size(group)
+        result /= torch.distributed.get_world_size(group)
 
     return result
 
diff --git a/pytorch_lightning/utilities/model_summary.py b/pytorch_lightning/utilities/model_summary.py
index 9c2690202..3789d6d62 100644
--- a/pytorch_lightning/utilities/model_summary.py
+++ b/pytorch_lightning/utilities/model_summary.py
@@ -415,7 +415,7 @@ def get_human_readable_count(number: int) -> str:
     num_groups = int(np.ceil(num_digits / 3))
     num_groups = min(num_groups, len(labels))  # don't abbreviate beyond trillions
     shift = -3 * (num_groups - 1)
-    number = number * (10 ** shift)
+    number *= (10 ** shift)
     index = num_groups - 1
     if index < 1 or number >= 100:
         return f"{int(number):,d} {labels[index]}"
```

</details>

**Comment by elfring** [comment]
>Would you mind sharing the motivation for this change other than PEP203 which you've shared in other issues?

I find that [the rationale of the Python enhancement proposal 203](https://www.python.org/dev/peps/pep-0203/#rationale "Proposal for augmented assignments") (from 2000-07-13) can indicate also motivation for another bit of collateral evolution. :thinking:


>I believe it won't work properly at some places …

Which of the concrete change suggestions (from which 34 source code transformations might look reasonable) would you find inappropriate (at the moment)?
* How will the chances evolve for related improvements?
* Would you eventually like to adjust the distinction where in-place modifications should be excluded?

**Comment by akihironitta** [comment]
> I find that [the rationale of the Python enhancement proposal 203](https://www.python.org/dev/peps/pep-0203/#rationale "Proposal for augmented assignments") (from 2000-07-13) can indicate also motivation for another bit of collateral evolution. :thinking:

The link you've provided is exactly the same as the one I mentioned in the previous reply (PEP203). Could you possibly fill up the "Motivation" section of this issue description above instead of just pasting the link?

> Which of the concrete change suggestions (from which 34 source code transformations might look reasonable) would you find inappropriate (at the moment)?

For example, this one:
https://github.com/PyTorchLightning/pytorch-lightning/blob/7a9a08c5d3ca4699fb439f691c40e1320b37507a/pytorch_lightning/loops/optimization/optimizer_loop.py#L90

> How will the chances evolve for related improvements?

I'm not sure what you're asking.

> Would you eventually like to adjust the distinction where in-place modifications should be excluded?

I'm not sure if I understand your question. We could probably replace some of the occurrences, but I see it as almost preference...

**Comment by elfring** [comment]
>Could you possibly fill up the "Motivation" section of this issue description above instead of just pasting the link?

Do you find the explanation by the Python enhancement proposal 203 (from 2000-07-13) insufficient? :thinking:


>For example, this one:

Why do you find the following source code transformation inappropriate (at the moment)? :thinking:

```diff
-            closure_loss = closure_loss / normalize
+            closure_loss /= normalize
```


>I'm not sure what you're asking.

:thought_balloon: Would another code review eventually be performed based on data processing results from the shown Perl command example?

**Comment by tchaton** [comment]
Dear @elfring,

I think `closure_loss /= normalize` would break with the latest PyTorch as it performs an in-place operation.

@ninginthecloud

**Comment by elfring** [comment]
* :thought_balloon: Are in-place modifications generally supported by all involved Python software libraries?
* Can run time characteristics eventually be influenced in desirable directions? :thinking:
